frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -625,105 +681,136 @@ pub mod pallet {
-        /// All stakers and developer for this contract will be paid out with single call.
-        /// claim is valid for all unclaimed eras but not longer than history_depth().
-        /// Any reward older than history_depth() will go to Treasury.
-        /// Any user can call this function.
-        #[pallet::weight(T::WeightInfo::claim(T::MaxNumberOfStakersPerContract::get() + 1))]
-        pub fn claim(
+        // TODO: do we need to add force methods or at least methods that allow others to claim for someone else?
+
+        /// Claim earned staker rewards for the oldest era.
+        #[pallet::weight(T::WeightInfo::claim_staker())]
+        pub fn claim_staker(
+            origin: OriginFor<T>,
+            contract_id: T::SmartContract,
+        ) -> DispatchResultWithPostInfo {
+            ensure!(!Self::pallet_disabled(), Error::<T>::Disabled);
+            let staker = ensure_signed(origin)?;
+
+            // Ensure we have something to claim
+            let mut staker_info = Self::staker_info(&staker, &contract_id);
+            let (era, staked) = staker_info.claim();
+            ensure!(staked > Zero::zero(), Error::<T>::NotStakedContract);
+
+            let dapp_info =
+                RegisteredDapps::<T>::get(&contract_id).ok_or(Error::<T>::NotOperatedContract)?; //should never happen
+            if let DAppState::Unregistered(unregister_era) = dapp_info.state {
+                ensure!(era < unregister_era, Error::<T>::NotOperatedContract);
+            }
+
+            let current_era = Self::current_era();
+
+            ensure!(era < current_era, Error::<T>::EraOutOfBounds);
+
+            let staking_info = Self::staking_info(&contract_id, era);
+            let reward_and_stake =
+                Self::era_reward_and_stake(era).ok_or(Error::<T>::UnknownEraReward)?;
+
+            let (_, stakers_reward) = Self::dev_stakers_split(
+                &staking_info,
+                &reward_and_stake,
+                &T::DeveloperRewardPercentage::get(),
+            );
+            let staker_reward = Perbill::from_rational(staked, staking_info.total) * stakers_reward;
+
+            // Withdraw reward funds from the dapps staking pot
+            let reward_imbalance = T::Currency::withdraw(
+                &Self::account_id(),
+                staker_reward,
+                WithdrawReasons::TRANSFER,
+                ExistenceRequirement::AllowDeath,
+            )?;
+
+            T::Currency::resolve_creating(&staker, reward_imbalance);
+            Self::update_staker_info(&staker, &contract_id, staker_info);
+
+            Self::deposit_event(Event::<T>::Reward(
+                staker.clone(),
+                contract_id.clone(),
+                era,
+                staker_reward,
+            ));
+
+            Ok(().into())
+        }
+
+        /// Claim earned dapp rewards for the specified era.
+        #[pallet::weight(T::WeightInfo::claim_dapp())]
+        pub fn claim_dapp(
```

Comment:
I'd still stick for a single claim call (stakers + dev) for the front-end but it is up to you and @bobo-k2